Easier depletion restart capability

GiudGiud opened the pull request:
This PR makes restarts of depletion calculations from the Python side easier, and adds a Python example to show how to do it.

One thing to consider : the atom densities in the summary.h5 file, written at only the first time step are not updated, which is why I load them from the depletion results instead. There are arguments for and against this choice, see below.  

**For using densities from depletion_results.h5 :** 
- we don't mind writing depletion_results.h5 at each time step, but we don't want to write the summary (large file) at each time step. Using depletion_results.h5, we can restart whenever we want too. 
- the depletion_results.h5 file has to be read anyway, to get the initial time of the restart, to concatenate results etc
- for complex integrators, we can use more than one time step to determine derivatives at the beginning of each restart, thus saving a few MC runs. This is not very convenient to do with the summary. 

**For using densities from statepoint.summary.geometry :** 
- less code on depletion side to handle restarts
- more consistent with how geometry and materials are loaded

With regards to volumes, they are currently not saved to the summary, as OpenMC doesn't need them to run. Doing that is a possible alternative to the new function (transfer_volumes) I created in results_list.py (should I move it to results.py btw).

With regards to how time evolves between simulations, an example below :
_Simulation 1 :_ 5 days in 1 day steps so transport calculations at days 0, 1, 2, 3, 4, 5
_Restart from simulation 1 :_ 5 days in 1 day steps so
- re-using day 5 reaction rates and depleting to day 6
- transport calculations at days 6, 7, 8, 9, 10

If you agree on how things are done on this PR, I will add unit tests and implement restart for the CECM integrator.

smharper commented on openmc/deplete/operator.py:
```diff
@@ -251,6 +273,40 @@ def _set_number_from_mat(self, mat):
+        geom_nuc = [x[0] for x in list(geom_nuc_densities.values())]
+
+        # Merge lists of nuclides
+        nuc_set = list(depl_nuc) + [n for n in geom_nuc if n not in depl_nuc]
```
Python has a built-in `set` object for stuff like this.  You should be able to replace this with something like,
```python
nuc_set = set(depl_nuc) | set(geom_nuc)
```

smharper commented on examples/python/pincell_depletion/restart_depletion.py:
```diff
@@ -0,0 +1,85 @@
+
+# Close statepoint and summary files to be able to write over them
+sp.summary._f.close()
+sp._f.close()
```
Attributes that begin with an underscore are meant to be for internal use only (i.e. users should never need to touch them).  To fix this, make a `close()` method for the `Summary` and `Statepoint` objects that calls the `_f.close()`

GiudGiud replied:
Ok. I found a __exit__ method for statepoints but I imagine that's still for internal use (called when program exits right ?). Anyway I added a method that calls __exit__, called close()

smharper commented on openmc/deplete/operator.py:
```diff
@@ -231,10 +245,18 @@ def _extract_number(self, local_mats, volume, nuclides):
-                self._set_number_from_mat(mat)
+        # Now extract and store the number densities
+        # From the geometry if no previous depletion results
+        if prev_res == None:
```
An explanation for these comments:  It's considered good practice in Python to do these `None` checks with `is` rather than `==`.  It might not be super relevant for this particular case, but it's a good habit to learn.  When you define a class, the `==` operator can be evaluated however you like and you might accidentally make `== None` evaluate to True, or there might be some obscure reason you do that on purpose.  In this case, we know that the specific Python object `None` has a special meaning so it's better to use the more restrictive comparison, `is`.

GiudGiud replied:
Thanks ! Learned some Python today

GiudGiud commented on openmc/deplete/results_list.py:
```diff
@@ -103,3 +103,18 @@ def get_eigenvalue(self):
+        for c in cells:
+            material = next(iter(cells[c].get_all_materials().values()))
+            if material.depletable == True:
+                material.volume = self[-1].volume[str(material.id)]
```
went for cell.get_all_materials().values() since mat_id can be obtained by material.id

GiudGiud commented:
Thanks for the review @smharper ! I addressed all of your comments.
@cjosey and @paulromano I'll wait on your approval on this way of doing depletion restarts before implementing unit tests and for CECM

paulromano commented on examples/python/pincell_depletion/restart_depletion.py:
```diff
@@ -0,0 +1,84 @@
+geometry = sp.summary.geometry
+
+# Close statepoint and summary files to be able to write over them
+sp.close()
```
The `StatePoint` class can be used as a context manager in Python, so you can change this to:
```Python
with openmc.StatePoint(statepoint) as sp:
    geometry = sp.summary.geometry
```
And then you don't need to call `sp.close()` at the end (it happens automagically)

paulromano commented on examples/python/pincell_depletion/restart_depletion.py:
```diff
@@ -0,0 +1,84 @@
+###############################################################################
+
+op = openmc.deplete.Operator(geometry, settings_file, chain_file, \
+                                                  previous_results)
```
No need for the explicit line continuation above; I would change to:
```Python
op = openmc.deplete.Operator(geometry, settings_file, chain_file,
                             previous_results)
```

paulromano commented on examples/python/pincell_depletion/restart_depletion.py:
```diff
@@ -0,0 +1,84 @@
+                                                                                                                                                                                                                                                                                                                                                                                                                ...
+# Plot eigenvalue as a function of time
+plt.figure()
+plt.plot(time/24/60/60, keff, label="K-effective")
```
`time/(24*60*60)`

Fun fact: the Python compiler performs [constant folding](https://en.wikipedia.org/wiki/Constant_folding) so that when your code is actually run, the interpreter only sees the single constant 86400. Here's a disassembler proof:
```Python
>>> import dis
>>> dis.dis('time/(24*60*60)')
  1           0 LOAD_NAME                0 (time)
              2 LOAD_CONST               3 (86400)
              4 BINARY_TRUE_DIVIDE
              6 RETURN_VALUE
```

paulromano commented on openmc/deplete/operator.py:
```diff
@@ -251,6 +276,42 @@ def _set_number_from_mat(self, mat):
+        # Get nuclide lists from geometry and depletion results
+        depl_nuc = prev_res[-1].nuc_to_ind.keys()
+        geom_nuc_densities = mat.get_nuclide_atom_densities()
+        geom_nuc = [x[0] for x in list(geom_nuc_densities.values())]
```
You can use a set comprehension here and then change `set(geom_nuc)` -> `geom_nuc` below. Also, there's no need for an explicit conversion to `list` since `values()` is iterable. So, it would be
```Python
geom_nuc = {x[0] for x in geom_nuc_densities.values()}
```

paulromano commented on openmc/statepoint.py:
```diff
@@ -153,6 +153,9 @@ def __exit__(self, *exc):
             self._summary._f.close()
 
+    def close(self):
+        self.__exit__()
```
Probably clearer to do the other way around: put the current block of code from `__exit__` here, and then have `__exit__` be:
```Python
def __exit__(self, *exc):
    self.close()
```
Also, this method needs a docstring.

GiudGiud replied:
I removed the method since I'm not using it anymore

paulromano commented on tests/dummy_operator.py:
```diff
@@ -18,6 +18,7 @@ class DummyOperator(TransportOperator):
 
     """
     def __init__(self):
+        self.prev_res = None
```
Get rid of pass below this

GiudGiud replied:
Was this pass just so that the method wouldn't be empty ?

GiudGiud commented:
Thanks for the review! Absolutely no problem on waiting.
I went ahead and added "unit" tests : 
- for transfer_volumes
- for restarting depletion
The second test is more of a regression test, kind of like test_deplete_predictor and test_deplete_cecm, in the same folder. 

I made combinations of initial / restart integrators, and it all works out except when the restart integration has more stages than the initial one. Then the formats for the data in depletion_results.h5 is too small for saving all the new data.
To solve this, I added a warning and save only the last stages that can fit.

smharper commented:
Everything looks good.  Merging.
